### ISSUE TITLE
Apply contribution guidelines

### ISSUE THREAD

**Issue by henriquemoody** [Respect/Validation#921]
TL;DR
- Make sure you're based on the `master` branch
- Refactor all rules as in #902 
- Write a comment with the rule you're  going to work on before sending a PR

***
It's been a MAJOR or MINOR version of the library has been released. There are enough things to release the version 2.0 but I need your help.

I wrote new [Contribution Guidelines](https://github.com/Respect/Validation/blob/master/CONTRIBUTING.md) and before I creating the tag 2.0.0 I would like to have all rules with the same standard.

Applying the Contribution Guidelines consist of:
* Add docblocks for classes, methods, and properties
* Add your `@author` tag to classes you changed
* Set properties as private
* Set classes (Rule and Exception) as final
* Add PHP 7.1 type hinting
* Make rule test extend RuleTestCase (when possible) 
* Make sure there is 1 (and 1 only) PHPT file for each rule testing only exceptions

I've done some already but there is still a lot to go, write a comment with the rule name before you start working on it, so two people don't work on the same thing!

Hope you can help!

## Rules

- [ ] ~AllOf~ (needs refactoring)
- [x] Alnum
- [x] Alpha
- [x] AlwaysInvalid
- [x] AlwaysValid
- [x] ~AnyOf~ (needs refactoring)
- [x] ArrayType
- [x] ArrayVal
- [x] Attribute
- [x] Base
- [x] Base64
- [x] Between
- [x] BoolType
- [x] BoolVal
- [x] Bsn
- [x] Call
- [x] CallableType
- [x] Callback
- [x] Charset
- [x] Cnh
- [x] Cnpj
- [x] Cntrl
- [x] Consonant
- [x] Contains
- [x] ContainsAny
- [x] Countable
- [x] CountryCode
- [x] Cpf
- [x] CreditCard
- [x] CurrencyCode
- [x] Date
- [x] DateTime
- [x] Digit
- [x] Directory
- [x] ~Domain~ (needs refactoring)
- [x] Each
- [x] Email
- [x] EndsWith
- [x] Equals
- [x] Equivalent
- [x] Even
- [x] Executable
- [x] Exists
- [x] Extension
- [x] Factor
- [x] FalseVal
- [x] Fibonacci
- [x] File
- [x] FilterVar
- [x] Finite
- [x] FloatType
- [x] FloatVal
- [x] Graph
- [x] GreaterThan
- [x] ~HexRgbColor~ (needs refactoring)
- [x] Identical
- [x] IdentityCard
- [x] Image
- [x] Imei
- [x] In
- [x] Infinite
- [x] Instance
- [x] IntType
- [x] IntVal
- [x] Ip
- [x] Isbn
- [x] IterableType
- [x] Json
- [x] ~Key~ (needs refactoring)
- [x] ~KeyNested~ (needs refactoring)
- [x] KeySet
- [x] KeyValue
- [x] LanguageCode
- [x] LeapDate
- [x] LeapYear
- [x] Length
- [x] LessThan
- [x] Lowercase
- [x] Luhn
- [x] MacAddress
- [x] Max
- [x] MaxAge
- [x] Mimetype
- [x] Min
- [x] MinAge
- [x] Multiple
- [x] Negative
- [x] NfeAccessKey
- [x] Nif
- [x] ~No~ (needs refactoring)
- [x] NoWhitespace
- [x] ~NoneOf~ (needs refactoring)
- [x] ~Not~ (needs refactoring)
- [x] NotBlank
- [x] ~NotEmpty~ (needs refactoring)
- [x] NotOptional
- [x] NullType
- [x] Nullable
- [x] Number
- [x] NumericVal
- [x] ObjectType
- [x] Odd
- [x] ~OneOf~ (needs refactoring)
- [x] ~Optional~ (needs refactoring)
- [x] PerfectSquare
- [x] Pesel
- [x] Phone
- [x] PhpLabel
- [x] Pis
- [x] Positive
- [x] ~PostalCode~ (needs refactoring)
- [x] PrimeNumber
- [x] Printable
- [x] Punct
- [x] Readable
- [x] Regex
- [x] ResourceType
- [x] ~Roman~ (needs refactoring)
- [x] ScalarVal
- [x] Sf
- [x] Size
- [x] Slug
- [x] Sorted
- [x] Space
- [x] StartsWith
- [x] StringType
- [x] StringVal
- [x] SubdivisionCode
- [x] Subset
- [x] SymbolicLink
- [x] Time
- [x] Tld
- [x] TrueVal
- [x] Type
- [x] Unique
- [x] Uploaded
- [x] Uppercase
- [x] Url
- [x] Uuid
- [x] Vatin
- [x] Version
- [x] VideoUrl
- [x] Vowel
- [x] When
- [x] Writable
- [x] ~Xdigit~ (needs refactoring)
- [x] Yes
- [x] ~Zend~ (needs refactoring)

Some rules are not on the list, but they are not *Easy Pick*.

**Comment by henriquemoody** [comment]
I'm getting the Between

**Comment by henriquemoody** [comment]
I'm getting the KeySet

**Comment by koficodes** [comment]
I'm getting  AllOf

**Comment by DaniloBenevides** [comment]
I'm getting Uppercase

**Comment by nickl-** [comment]
Well done! whoop whoop!!

Unfortunately I'm not quite free right now... knee deep in java, but will try and wriggle my way out before this is done.

**Comment by koficodes** [comment]
while working on AllOf , I realized it is being extended by other Rules, hence setting the class as final is not feasible.
also adding type hinting will cause a cascade of incompatible interfaces and function signatures.
I suggest the following task should be tackled in one phase and the other two should be tackled in another phase with much diligence.

- Add docblocks for classes, methods, and properties
- Add @author tag to classes (script that helps doing that: https://gist.github.com/henriquemoody/d4409df72a86dd07b9a794635ce0995a)
- Set properties as final (if it is not extended by other classes)
- Make rule test extend RuleTestCase (when possible)
- Make sure there is 1 (and 1 only) PHPT file for each rule testing only exceptions

-----------------------------

- Set classes (Rule and Exception) as final (if it is extended by other classes)
- Add PHP 7 type hinting
-

**Comment by henriquemoody** [comment]
Some rules require more than applying the guidelines. I updated the list of Rules in the description of this issue, mentioning only the ones that are easy to work on, I'll pick up myself because of the more structural changes. 

If you can pick other rule that will be simpler that would be the cool. If you have something already done already in the "AllOf" rule, you can just send it as it, and I'll pick it up from there you left! But I think we should try to apply the complete list at once, there are too many rules to do into more steps, IMO. 

Thanks for your feedback, @koficodes! 🐼

**Comment by koficodes** [comment]
Well noted. will pick another rule for now. Thanks for the quick response. :+1:

**Comment by williamespindola** [comment]
I'm getting the ArrayType

**Comment by koficodes** [comment]
I'm getting Even

**Comment by DaniloBenevides** [comment]
Im getting Lowercase

**Comment by williamespindola** [comment]
Guys I found a bug on Contains rule during the work. I would send a patch but looking at the [history of file](https://github.com/Respect/Validation/commits/master/library/Rules/Contains.php) this problem exists for a long time ago. I think that we can continue this work and I will join the fix with apply of this issue. What you think about that?
About the problem. The Contains rule do not check if an item of array has no identical value because it checks if is array before execute the verification with no identical. This will work `v:contains('foo')->check('fool')` but this don't `v:contains('foo')->check(['fool', 'bar'])`

**Comment by henriquemoody** [comment]
@williamespindola while applying the contribution guidelines I've found a few bugs myself!

However, if I understood what you said correctly that's the expected behaviour.

If you're searching in a string, you need a match **in** the string:

```php
v:contains('foo')->validate('fool'); // true
v:contains('foo')->validate('fol'); // false
```

If you're searching in an array, you need a  match **in** the array:

```php
v:contains('foo')->validate(['foo', 'bar'); // true
v:contains('foo')->validate(['fol', 'bar'); // false
```

It's inverted behaviour of "In".

**Comment by henriquemoody** [comment]
I'm getting IterableType

**Comment by williamespindola** [comment]
@henriquemoody I don't know if I get it wrong. But the Contais rule have [a flag on constructor](https://github.com/Respect/Validation/blob/master/library/Rules/Contains.php#L21) to match value identical or not. Then if we passe true it will sought for exactly `foo` but whether we pass false or nothing this will sought for a similar e.g. `foo == fool` take a look https://3v4l.org/mYGVe

**Comment by williamespindola** [comment]
I'm getting Countable

**Comment by williamespindola** [comment]
I'm getting Cpf

**Comment by williamespindola** [comment]
I'm getting AlwaysInvalid and AlwaysValid

**Comment by williamespindola** [comment]
I'm working on ArrayVal, Base64, Base, BoolType and BoolVal

**Comment by DaniloBenevides** [comment]
I am getting FloatVal and FloatType

**Comment by Ismael-Elias** [comment]
i'm getting Positive

**Comment by IsmaelElias** [comment]
I'm getting Negative

**Comment by DaniloBenevides** [comment]
I am getting Odd

**Comment by IsmaelElias** [comment]
I'm getting PrimeNumber

**Comment by koficodes** [comment]
I'm getting email

**Comment by DaniloBenevides** [comment]
I'm getting NoWhitespace

**Comment by nickl-** [comment]
31% completed... this is taking for ever =(

Good job so far guys!

**Comment by williamespindola** [comment]
working on Bsn, Callback, Charset, Cnh, Cnpj

**Comment by DaniloBenevides** [comment]
I'm getting PerfectSquare.

**Comment by williamespindola** [comment]
I'm getting CountryCode, CreditCard, CurrencyCode, Directory and Each

**Comment by IsmaelElias** [comment]
I'm getting Number

**Comment by koficodes** [comment]
I'm getting :

-   Tld
-   TrueVal
-   Type
-   Unique
-   Uploaded

**Comment by henriquemoody** [comment]
There is a new version of the script: https://gist.github.com/henriquemoody/d4409df72a86dd07b9a794635ce0995a

**Comment by williamespindola** [comment]
I'm working on: EndsWith, Executable, Exists, Extension, Factor, FalseVal, Fibonacci, File, Finite

**Comment by koficodes** [comment]
I'm working on: 
-   Not
-   NotBlank
-   NotEmpty
-   NotOptional
-   Number
-   NumericVal

**Comment by nickl-** [comment]
52% Whoop whoop!! It's all downhill from here =)

Go-go-go!!

**Comment by DaniloBenevides** [comment]
I'm gettin intVal and Multiple

**Comment by nickl-** [comment]
Only 54% now, what's happening? We were doing so well!

I saw @henriquemoody agreed to release v1.1.19 is this the reason progress is impacted?

Come on guys almost there, lets get this done!

**Comment by nickl-** [comment]
I must've missed the memo... has this development now stopped?

**Comment by henriquemoody** [comment]
Not at all! I did spend quite some time on the version 1.1.19, there were quite a few bugs. In retrospect, that was a bad decision. Now I'm struggling for a few days with the merge of the branch 1.1 into master. However, it should not stop! As soon as I merge the 1.0 I'm right back into the 2.0!

How about you, @koficodes, @DaniloBenevides, and @williamespindola?

And thanks for pushing it forward, @nickl-!

**Comment by williamespindola** [comment]
[I'll be back](https://media.giphy.com/media/JDKxRN0Bvmm2c/giphy.gif) Just solving some issue in my work

**Comment by DaniloBenevides** [comment]
I'm getting Image and Imei

**Comment by nickl-** [comment]
It seems our progress is still stuck at 54% completion. 

Is there any obstacles preventing this from moving forward which we could move out of the way or is this purely due to availability constraints?

How about a commitment to get 55%?
@DaniloBenevides are you done with Imei and Image yet?
@williamespindola has the Sky Labs mainframe authorised your return yet? =)

Lets get this wheel turning again, what do you say?

**Comment by nickl-** [comment]
Added tick to Multiple rule #1051 ...no progress was really getting me down =)

**Comment by DaniloBenevides** [comment]
I'm working on Image Rule , will try to finish till weekend. 
I think i can try 5 rules per week from next monday. :smile:

**Comment by nickl-** [comment]
@DaniloBenevides Go get them tiger! =)

gl

**Comment by DaniloBenevides** [comment]
I'm getting Infinite , Instance and In

**Comment by DaniloBenevides** [comment]
I'm getting Ip , Json, KeyValue, LanguageCode, LeapDate

**Comment by nickl-** [comment]
58% complete... steady as she goes.

Nice! =)

**Comment by IsmaelElias** [comment]
I'm working on: 
- Sf
- Size  
- Slug
- Sorted
- StartsWith

**Comment by danilosilva87** [comment]
I'm working Length

**Comment by danilosilva87** [comment]
@henriquemoody  I think the `IN` rule has already been made, but it has not been checked.

**Comment by danilosilva87** [comment]
I'm working on: 
* NumericVal
* StringVal

**Comment by danilosilva87** [comment]
I'm working on:
* Pesel
* PhpLabel
* Pis

**Comment by danilosilva87** [comment]
I'm working on:
* Readable
* Regex

**Comment by henriquemoody** [comment]
I removed some rules from the list. There are some rules that I still have to figure out what to do with them:

- [ ] AllOf
- [ ] Alpha
- [ ] AnyOf
- [ ] Cntrl
- [ ] Consonant
- [ ] Digit
- [ ] Domain
- [ ] Graph
- [ ] HexRgbColor
- [ ] Key
- [ ] KeyNested
- [ ] No
- [ ] NoneOf
- [ ] Not
- [ ] NotEmpty
- [ ] OneOf
- [ ] Optional
- [ ] Phone
- [ ] PostalCode
- [ ] Punct
- [ ] Regex
- [ ] Roman
- [ ] Space
- [ ] Uuid
- [ ] Vowel
- [ ] Xdigit
- [ ] Yes

I'm picking these ones up. I will take what was already one in #1101 and #1102.

**Comment by danilosilva87** [comment]
I'm working on:
* NotBlank
* NotOptional

**Comment by danilosilva87** [comment]
I'm working on LeapYear

**Comment by danilosilva87** [comment]
I'm working on:
* Luhn
* MacAddress
* Mimetype

**Comment by nickl-** [comment]
Was-Wah-Wee-Waa!!! 77% is very nice!!!

=)

**Comment by CameronHall** [comment]
I'm working on Yes

**Comment by danilosilva87** [comment]
> I'm working on: EndsWith, Executable, Exists, Extension, Factor, FalseVal, Fibonacci, File, Finite

@williamespindola as we talked in the pub, I'll get the rules  `Extension, Factor, FalseVal, Fibonacci, File, Finite`, ok?

**Comment by danilosilva87** [comment]
I'm working on:
* Version
* VideoUrl
* When
* Writable
* Zend

**Comment by leonampd** [comment]
I'm working on:

- NfeAccessKey
- Phone
- Punct

**Comment by GusAntoniassi** [comment]
Hello, I will pick up `SymbolicLink`. From what I gathered, it's the only one not being worked on, and the list of available options (including the ones being worked on) would be:
- [ ] Size
- [ ] Size
- [ ] Size
- [ ] Size
- [ ] Size
- [ ] Size

**Comment by henriquemoody** [comment]
Happy to hear you would like to contribue, @GusAntoniassi! 

The rules that still need to be updated but that are _easy to pick_ are:

- Cntrl
- Consonant
- Graph
- KeyValue
- NfeAccessKey
- Phone
- Punct
- Size
- Slug
- Sorted
- Space
- StartsWith
- SymbolicLink
- Vowel

However, there are also other rules that needs some refactoring, in which each case needs some speciall discussion:

- AllOf
- AnyOf
- Domain
- HexRgbColor
- Ip
- Key
- KeyNested
- No
- NoneOf
- Not
- NotEmpty
- OneOf
- Optional
- PostalCode
- Regex
- Roman
- Xdigit
- Zend

Feel free to pick any of them!

**Comment by henriquemoody** [comment]
It's been quite a ride! 

The only rule that is missing now is `AllOf`, the rest is pretty much done, which means that it won't take long to release 2.0! I would need to create some good documentation about the release before doing it, but we are very close!

**Comment by mazentouati** [comment]
what's wrong with `AllOf` ? is there anyone working on it ?

**Comment by henriquemoody** [comment]
The most important thing is that it becomes `final`. What could be nice was to remove the `addRule` from `AbstractComposite`. It's a bit complicated because `Validator` itself extends it -- which for me justifies having an abstract class. I've started to work on that already, but I will only have time to keep working on it next week, so if anyone would like to work on it, feel free, just let me know!

**Comment by henriquemoody** [comment]
This is being shipped already!
